lnwallet/commitment.go

```diff
@@ -0,0 +1,711 @@
+		)
+	} else {
+		return DeriveStateHintObfuscator(
+			state.RemoteChanCfg.PaymentBasePoint.PubKey,
+			state.LocalChanCfg.PaymentBasePoint.PubKey,
+		)
+	}
+}
+
+// commitmentView is the final commitment created from evaluating an HTLC view
+// at a given height.
+type commitmentView struct {
+	// txn is the final, unsigned commitment transaction for this view.
+	txn *wire.MsgTx
+
+	// fee is the total fee of the commitment transaction.
+	fee btcutil.Amount
+
+	// ourBalance|theirBalance are the balances of this commitment *after*
+	// subtracting commitment fees, but NOT anchor outputs. This can be
+	// different than the balances before creating the commitment
+	// transaction as one party must pay the commitment fee.
+	ourBalance   lnwire.MilliSatoshi
+	theirBalance lnwire.MilliSatoshi
+}
+
+// createCommitmentTx generates the unsigned commitment transaction for a
+// commitment view and assigns to txn field. The passed in balances should be
+// balances *before* subtracting any commitment fees.
+func (cb *CommitmentBuilder) createCommitmentTx(ourBalance,
+	theirBalance lnwire.MilliSatoshi, isOurs bool,
+	feePerKw chainfee.SatPerKWeight, height uint64,
+	filteredHTLCView *htlcView,
+	keyRing *CommitmentKeyRing) (*commitmentView, error) {
+
+	dustLimit := cb.chanState.LocalChanCfg.DustLimit
+	if !isOurs {
+		dustLimit = cb.chanState.RemoteChanCfg.DustLimit
+	}
+
+	numHTLCs := int64(0)
+	for _, htlc := range filteredHTLCView.ourUpdates {
+		if htlcIsDust(false, isOurs, feePerKw,
+			htlc.Amount.ToSatoshis(), dustLimit) {
+
+			continue
+		}
+
+		numHTLCs++
+	}
+	for _, htlc := range filteredHTLCView.theirUpdates {
+		if htlcIsDust(true, isOurs, feePerKw,
+			htlc.Amount.ToSatoshis(), dustLimit) {
+
+			continue
+		}
+
+		numHTLCs++
+	}
+
+	// Next, we'll calculate the fee for the commitment transaction based
+	// on its total weight. Once we have the total weight, we'll multiply
+	// by the current fee-per-kw, then divide by 1000 to get the proper
+	// fee.
+	totalCommitWeight := CommitWeight(cb.chanState.ChanType) +
+		input.HTLCWeight*numHTLCs
+
+	// With the weight known, we can now calculate the commitment fee,
+	// ensuring that we account for any dust outputs trimmed above.
+	commitFee := feePerKw.FeeForWeight(totalCommitWeight)
+	commitFeeMSat := lnwire.NewMSatFromSatoshis(commitFee)
+
+	// Currently, within the protocol, the initiator always pays the fees.
+	// So we'll subtract the fee amount from the balance of the current
+	// initiator. If the initiator is unable to pay the fee fully, then
+	// their entire output is consumed.
+	switch {
+	case cb.chanState.IsInitiator && commitFee > ourBalance.ToSatoshis():
+		ourBalance = 0
+
+	case cb.chanState.IsInitiator:
+		ourBalance -= commitFeeMSat
+
+	case !cb.chanState.IsInitiator && commitFee > theirBalance.ToSatoshis():
+		theirBalance = 0
+
+	case !cb.chanState.IsInitiator:
+		theirBalance -= commitFeeMSat
+	}
+
+	// If the commitment has anchors, we must also subtract the anchor
```

Comment:
This is a bit awkward, but is a result of that the balances we store in the db has had their commit fee subtracted. should we do a db migration to change this?